lib/recursive-methods.rb

```diff
@@ -1,49 +1,88 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    if n == 0
+        return 1
+    elsif n < 0
+        raise ArgumentError
+    else
+        return n * factorial(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length <= 1
+        return s
+    else
+        return reverse(s[1..-1]) + s[0]
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# # Time complexity: O(1)
+# # Space complexity: O(1)
 def reverse_inplace(s)
```

Comment:
⚠️ How could you reverse an entire string in O(1) time?

Also you're using the exact method I told you not to use.